browser/lib/util/bufferutils.js

```diff
@@ -90,19 +90,19 @@ var BufferUtils = (function() {
 	BufferUtils.base64Decode = function(str) {
 		if(ArrayBuffer)
 			return base64ToArrayBuffer(str);
-
-		if(CryptoJS)
-			return CryptoJS.enc.Base64.parse(str);
+		return CryptoJS.enc.Base64.parse(str);
 	};
 
 	BufferUtils.utf8Encode = function(string) {
-		if(CryptoJS)
-			return CryptoJS.enc.Utf8.parse(string);
+		return CryptoJS.enc.Utf8.parse(string);
 	};
 
 	BufferUtils.utf8Decode = function(buf) {
-		if(CryptoJS)
+		if(isArrayBuffer(buf))
+			buf = BufferUtils.toWordArray(buf) // CryptoJS only works with WordArrays
+		if(isWordArray(buf))
 			return CryptoJS.enc.Utf8.stringify(buf);
+		throw new Error("Expected input of utf8Decode to be a buffer");
```

Comment:
Ideally when the exception is caught in decode(), the resultant error will identity the encoding component that failed (ie `utf-8`), and will include the error message from the root error. I don't think it needs to be quite as verbose as what you suggest there.

I think 400 is the most appropriate `statusCode`, and probably a new `code` `40013` `Invalid message data or encoding`

I understand that the `ErrorInfo` fields don't make it into the final error given to the client but I'd still prefer that it's an ErrorInfo. We might in the future have a `cause` field, for example, where outer errors wrap the underlying error.